bigchaindb/models.py

```diff
@@ -0,0 +1,211 @@
+        if transactions is not None and not isinstance(transactions, list):
+            raise TypeError('`transactions` must be a list instance or None')
+        elif transactions is None:
+            self.transactions = []
+        else:
+            self.transactions = transactions
+
+        if voters is not None and not isinstance(voters, list):
+            raise TypeError('`voters` must be a list instance or None')
+        elif transactions is None:
+            self.voters = []
+        else:
+            self.voters = voters
+
+        if timestamp is not None:
+            self.timestamp = timestamp
+        else:
+            self.timestamp = gen_timestamp()
+
+        self.node_pubkey = node_pubkey
+        self.signature = signature
+
+    def __eq__(self, other):
+        try:
+            other = other.to_dict()
+        except AttributeError:
+            return False
+        return self.to_dict() == other
+
+    def validate(self, bigchain):
+        """Validate a block.
+
+        Args:
+            bigchain (Bigchain): an instantiated bigchaindb.Bigchain object.
+
+        Returns:
+            block (Block): The block as a `Block` object if it is valid.
+                           Else it raises an appropriate exception describing
+                           the reason of invalidity.
+
+        Raises:
+            OperationError: if a non-federation node signed the block.
+        """
+
+        # First, make sure this node hasn't already voted on this block
+        if bigchain.has_previous_vote(self.id, self.voters):
+            return self
+
+        # Check if the block was created by a federation node
+        possible_voters = (bigchain.nodes_except_me + [bigchain.me])
+        if self.node_pubkey not in possible_voters:
+            raise OperationError('Only federation nodes can create blocks')
+
+        if not self.is_signature_valid():
+            raise InvalidSignature('Block signature invalid')
+
+        # Finally: Tentative assumption that every blockchain will want to
+        # validate all transactions in each block
+        for tx in self.transactions:
+            # NOTE: If a transaction is not valid, `is_valid` will throw an
+            #       an exception and block validation will be canceled.
+            bigchain.validate_transaction(tx)
+
+        return self
+
+    def sign(self, private_key):
+        block_body = self.to_dict()
+        block_serialized = serialize(block_body['block'])
+        private_key = SigningKey(private_key)
+        self.signature = private_key.sign(block_serialized)
+        return self
+
+    def is_signature_valid(self):
+        block = self.to_dict()['block']
+        block_serialized = serialize(block)
+        verifying_key = VerifyingKey(block['node_pubkey'])
+        try:
+            return verifying_key.verify(block_serialized, self.signature)
+        # TODO FOR CC: Shouldn't throw `AttributeError` or `ValueError` but
```

Comment:
I've created a ticket for this in the cryptoconditions repo: https://github.com/bigchaindb/cryptoconditions/issues/27